StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/IntentStatusPollerTest.swift

```diff
@@ -0,0 +1,115 @@
+//
+//  IntentStatusPollerTest.swift
+//  StripePaymentSheetTests
+//
+//  Created by Nick Porter on 9/14/23.
+//
+
+@testable import StripePaymentSheet
+import StripePaymentsTestUtils
+import XCTest
+
+class IntentStatusPollerTest: XCTestCase {
+    var sut: IntentStatusPoller!
+    var mockIntentRetriever: MockPaymentIntentRetriever!
+    var mockDelegate: MockIntentStatusPollerDelegate!
+    var intentRetrieverExpectation: XCTestExpectation!
+    var delegateExpectation: XCTestExpectation!
+
+    override func setUp() {
+        super.setUp()
+        mockIntentRetriever = MockPaymentIntentRetriever()
+        mockDelegate = MockIntentStatusPollerDelegate()
+        sut = IntentStatusPoller(retryInterval: 1.0, intentRetriever: mockIntentRetriever, clientSecret: "test_client_secret")
+        sut.delegate = mockDelegate
+    }
+
+    override func tearDown() {
+        sut = nil
+        mockIntentRetriever = nil
+        mockDelegate = nil
+        super.tearDown()
+    }
+
+    func setExpectations(apiExpectedCount: Int, delegateExpectedCount: Int) {
+        intentRetrieverExpectation = XCTestExpectation()
+        delegateExpectation = XCTestExpectation()
+        delegateExpectation.assertForOverFulfill = true
+        intentRetrieverExpectation.expectedFulfillmentCount = apiExpectedCount
+        delegateExpectation.expectedFulfillmentCount = delegateExpectedCount
+
+        mockIntentRetriever.expectation =  intentRetrieverExpectation
+        mockDelegate.expectation = delegateExpectation
+    }
+
+    func testPolling_beginSuspendBegin() {
+        // Poll 3 times
+        // Should call retrievePaymentIntent 3 times
+        // Delegate should be notified on the first poll but not subsequent polls
+        setExpectations(apiExpectedCount: 3, delegateExpectedCount: 1)
+        mockIntentRetriever.mockedStatus = .requiresPaymentMethod
+
+        sut.beginPolling()
+        XCTAssertTrue(sut.isPolling)
+
+        wait(for: [intentRetrieverExpectation, delegateExpectation], timeout: 5.0)
+        XCTAssertEqual(mockDelegate.latestPaymentIntent?.status, .requiresPaymentMethod)
+
+        sut.suspendPolling() // We should no longer notify the delegate or call the API
+        XCTAssertFalse(sut.isPolling)
+
+        // Make the API client return succeeded, then the delegate should NOT be notified, and API should NOT be called due to suspended polling
+        setExpectations(apiExpectedCount: 1, delegateExpectedCount: 1)
+        intentRetrieverExpectation.isInverted = true // expectations should not be fufilled since polling is suspended
+        delegateExpectation.isInverted = true
+        mockIntentRetriever.mockedStatus = .succeeded
+
+        wait(for: [intentRetrieverExpectation, delegateExpectation], timeout: 5.0)
+        // delegate should not have been notified of succeeded since polling was suspended
+        XCTAssertEqual(mockDelegate.latestPaymentIntent?.status, .requiresPaymentMethod)
```

Comment:
Does `wait` here wait 5 seconds for the expectations, or does it just immediately succeed?

b/c we should be waiting like `interval + buffer` seconds before validating that the poller didn't make network requests etc.

If we don't want to fiddle with expectations, I think simply waiting and then validating that `lastPaymentIntent.status` didn't update would work.

Reply:
In a functional sense they are pretty much succeeding right away since now our retry interval is 0.1 seconds. If the expectations aren't being fulfilled it will wait here for the full 5 seconds. But since they are being fulfilled so fast it's not really waiting. I prefer this approach over `wait(1)` or something as it waits the minimum amount of time, and just feels a bit better to me.

Reply:
If we have a bug in our poller where it continues to poll after you call `suspendPolling`, would this catch it?  If I'm understanding, it wouldn't - we _immediately_ check that the poller didn't make a network request or update the PI and succeed, instead of waiting until the next `interval`.

Reply:
Oop, I was looking at the wrong `wait`. I was referencing the one on line 66.

The `wait` on line 58, for the inverted expectations does wait the full five seconds, ensuring they are not fulfilled. Which ensures we aren't making any API requests or invoking the delegate while suspended.

I added some print statements around it
```
started waiting 1694727167.03133
ended waiting 1694727172.035596
```

Reply:
Ah good - then we should change the timeout to be something like `interval*2` (x2 to allow buffer for the timer).

Reply:
Done!